src/randoms.jl

```diff
@@ -1,43 +1,130 @@
+    return GaussianUnitary(disp, symp)
+end
+function _randunitary_fields(nmodes::N, passive) where {N<:Int}
+    disp = rand(2*nmodes)
+    symp = randsymplectic(nmodes, passive = passive)
+    return disp, symp
+end
 
 """
     randchannel([Td=Vector{Float64}, Tt=Matrix{Float64},] nmodes<:Int)
 
 Calculate a random Gaussian channel.
 """
 function randchannel(::Type{Td}, ::Type{Tt}, nmodes::N) where {Td,Tt,N<:Int}
-    disp = randn(2*nmodes)
-    transform = randn(2*nmodes, 2*nmodes)
-    randmat = randn(2*nmodes, 2*nmodes)
-    # positive-definite condition
-    noise = randmat' * randmat
+    disp, transform, noise = _randchannel(nmodes)
     return GaussianChannel(Td(disp), Tt(transform), Tt(noise), nmodes)
 end
 randchannel(::Type{T}, nmodes::N) where {T,N<:Int} = randchannel(T,T,nmodes)
 function randchannel(nmodes::N) where {N<:Int}
-    disp = randn(2*nmodes)
-    transform = randn(2*nmodes, 2*nmodes)
-    randmat = randn(2*nmodes, 2*nmodes)
-    # positive-definite condition
-    noise = randmat' * randmat
+    disp, transform, noise = _randchannel(nmodes)
     return GaussianChannel(disp, transform, noise, nmodes)
+end
+function _randchannel(nmodes::N) where {N<:Int}
+    disp = randn(2*nmodes)
+    # generate symplectic matrix describing the evolution of the system with N modes
+    # and environment with 2N modes
+    m = 2*nmodes # number of modes in environment
+    symp = randsymplectic(nmodes + m)
+    transform, B = symp[1:2*nmodes, 1:2*nmodes], @view(symp[1:2*nmodes, 2*nmodes+1:2*(nmodes + m)])
+    # generate random covariance matrix of environment
+    envsymp = randsymplectic(m)
+    envcovar = envsymp * envsymp'
+    # create buffers for matrix multiplication
+    buf = zeros(2*m, 2*nmodes)
+    # generate noise matrix from evolution of environment
+    noise = zeros(2*nmodes, 2*nmodes)
+    mul!(noise, B, mul!(buf, envcovar, B'))
+    return disp, transform, noise
+end
+
+"""
+    randsymplectic([T=Matrix{Float64},] nmodes<:Int, passive=false)
+
+Calculate a random symplectic matrix of size `2*nmodes x 2*nmodes`.
+"""
+function randsymplectic(nmodes::N; passive = false) where {N<:Int}
+    # generate random orthogonal symplectic matrix
+    O = _rand_orthogonal_symplectic(nmodes)
+    if passive
+        return O
+    end
+    # instead generate random active symplectic matrix
+    O′ = _rand_orthogonal_symplectic(nmodes)
+    # direct sum of symplectic matrices for single-mode squeeze transformations
+    rs = rand(nmodes)
+    squeezes = diagm(vcat(exp.(-rs), exp.(rs)))
+    return O * squeezes * O′
+end
+function randsymplectic(::Type{T}, nmodes::N; passive = false) where {T, N<:Int} 
+    symp = randsymplectic(nmodes, passive)
+    return T(symp)
+end
+
+# Generates unitary matrix randomly distributed over Haar measure;
+# see https://arxiv.org/abs/math-ph/0609050 for algorithm.
+# This approach is much faster than using rand(Haar(2), nmodes) from RandomMatrices.jl
+function _rand_unitary(nmodes::N) where {N<:Int}
```

Comment:
Is this so much faster that it might be worth changing the implementation in QuantumOpticsBase.jl to match?

Reply:
Yes, I would say so:
```
julia> @benchmark rand(Haar(2), 20)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  22.625 μs …   8.383 ms  ┊ GC (min … max):  0.00% … 99.19%
 Time  (median):     27.250 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   31.482 μs ± 151.002 μs  ┊ GC (mean ± σ):  12.35% ±  2.79%

                ▂▄▂▆█▆▄▁                                        
  ▁▃▂▂▂▂▂▂▂▂▂▃▄▆█████████▇▅▄▄▃▄▄▃▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  22.6 μs         Histogram: frequency by time         37.2 μs <

 Memory estimate: 58.45 KiB, allocs estimate: 41.

julia> @benchmark Gabs._rand_unitary(10)
BenchmarkTools.Trial: 10000 samples with 5 evaluations.
 Range (min … max):  6.917 μs …  1.210 ms  ┊ GC (min … max): 0.00% … 98.92%
 Time  (median):     7.342 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.082 μs ± 20.767 μs  ┊ GC (mean ± σ):  6.53% ±  3.42%

   ▅▇   ▁▁▄█▆▇                                                
  ▃██▇█▆███████▇▄▂▃▂▂▂▃▃▃▄▄▆▆▆▆▅▆▅▆▄▅▃▃▄▃▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁ ▃
  6.92 μs        Histogram: frequency by time        9.06 μs <

 Memory estimate: 14.92 KiB, allocs estimate: 19.
```
Plus, RandomMatrices needs a lot of love on the maintenance side, so I'm hesitant to add it as a dependency for just one feature.

Reply:
Fair, something to keep in mind though if QuantumOpticsBase ever ends up as a dependency for Gabs

Reply:
I agree. The QO maintainers might be interested in removing RandomMatrices as a dependency, I know they are working in reducing the slow startup time (you may have already seen this issue but I am putting it here anyways: https://github.com/qojulia/QuantumOptics.jl/issues/417).